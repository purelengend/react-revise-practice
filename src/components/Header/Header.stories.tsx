import type { Meta, StoryObj } from "@storybook/react";

// Components
import Header from ".";

const meta: Meta<typeof Header> = {
  title: "Components/Header",
  component: Header,
  decorators: [
    (Story) => (
      <div style={{ width: "100vh" }}>
        <Story />
      </div>
    ),
  ],
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof Header>;

export const Primary: Story = {};
