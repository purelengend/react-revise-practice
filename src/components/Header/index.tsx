import { HamburgerIcon } from "@chakra-ui/icons";
import { Center, HStack, IconButton } from "@chakra-ui/react";
import { useCallback } from "react";
import { useSearchParams } from "react-router-dom";

// Components
import { NotificationIcon } from "../common/Icons";
import { SearchForm, SearchFormInput } from "../common";
import { QUERY_PARAMS } from "@/constants";

export type HeaderProps = {
  onToggleSidebar: () => void;
};
const Header = ({ onToggleSidebar }: HeaderProps) => {
  // const { refetchAllStudents } = useStudentPagination();

  const [searchParams, setSearchParams] = useSearchParams();

  const name = searchParams.get(QUERY_PARAMS.NAME) || "";

  const handleSearchSubmit = useCallback(
    ({ searchValue }: SearchFormInput) => {
      // Reset page index when searching
      searchParams.delete(QUERY_PARAMS.PAGE);

      searchValue
        ? searchParams.set(QUERY_PARAMS.NAME, searchValue)
        : searchParams.delete(QUERY_PARAMS.NAME);

      setSearchParams(searchParams);
    },
    [searchParams, setSearchParams],
  );

  return (
    <Center px={4} justifyContent="space-between" h={15}>
      <IconButton
        aria-label="Toggle sidebar"
        icon={<HamburgerIcon boxSize={8} />}
        onClick={onToggleSidebar}
      />
      <HStack gap={6} mr={{ md: 12 }}>
        <SearchForm searchParam={name} onSubmit={handleSearchSubmit} />
        <NotificationIcon />
      </HStack>
    </Center>
  );
};

export default Header;
