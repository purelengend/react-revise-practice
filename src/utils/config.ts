// Check import.meta.env is undefined in test environment
export const required = (key: string, defaultValue = undefined) => {
  if (import.meta.env) return import.meta.env[key];
  return defaultValue;
};
