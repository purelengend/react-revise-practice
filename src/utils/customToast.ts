import { TOAST_DEFAULT_OPTIONS } from "@/constants";

import { AlertStatus, UseToastOptions } from "@chakra-ui/react";
export const customToast = (
  title: string,
  description: string,
  status: AlertStatus,
): UseToastOptions => {
  if (status === "success") {
    return {
      ...TOAST_DEFAULT_OPTIONS.SUCCESS,
      title,
      description,
    };
  }

  return {
    ...TOAST_DEFAULT_OPTIONS.ERROR,
    title,
    description,
  };
};
