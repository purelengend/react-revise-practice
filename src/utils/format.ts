export const formatPhoneNumber = (phoneNumber: string) => {
  // Check if the phone number is valid (10 digits)
  if (!/^\d{10}$/.test(phoneNumber)) {
    throw new Error("Invalid phone number. Must be a 10-digit number.");
  }

  // Format the phone number
  const areaCode = phoneNumber.slice(0, 3);
  const middlePart = phoneNumber.slice(3, 6);
  const lastPart = phoneNumber.slice(6, 10);

  return `${areaCode}-${middlePart}-${lastPart}`;
};
