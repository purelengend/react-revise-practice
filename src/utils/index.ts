export * from "./test-utils";
export * from "./config";
export * from "./customToast";
