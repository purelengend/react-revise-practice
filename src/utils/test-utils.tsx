/* eslint-disable react-refresh/only-export-components */
import { RenderOptions, render } from "@testing-library/react";
import { ReactElement, ReactNode } from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { MemoryRouter } from "react-router-dom";

const queryClient = new QueryClient({});
export const AllTheProviders = (props: { children: ReactNode }) => {
  return (
    <MemoryRouter initialEntries={["/"]}>
      <QueryClientProvider client={queryClient}>
        {props.children}
      </QueryClientProvider>
    </MemoryRouter>
  );
};

export const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
