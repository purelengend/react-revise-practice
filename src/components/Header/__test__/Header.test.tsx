import { act } from "react";
import { customRender } from "@/utils";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

// Components
import Header from "..";

describe("Header test cases", () => {
  const mockToggle = jest.fn();

  const setup = () => customRender(<Header onToggleSidebar={mockToggle} />);
  it("should render correctly", async () => {
    const { container } = await act(() => setup());

    expect(container).toMatchSnapshot();
  });

  it("should invoke onChange function when typing", async () => {
    setup();

    const searchInput = screen.getByRole("textbox");

    await userEvent.type(searchInput, "mock search");
  });
});
