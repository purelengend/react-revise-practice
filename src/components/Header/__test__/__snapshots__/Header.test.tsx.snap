// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header test cases should render correctly 1`] = `
<div>
  <div
    class="css-gcgo19"
  >
    <button
      aria-label="Toggle sidebar"
      class="chakra-button css-p7mkev"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-r8yn95"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      class="chakra-stack css-1fvwrdw"
    >
      <form
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          class="chakra-input css-sh2byc"
          name="searchValue"
          placeholder="Search..."
          value=""
        />
        <div
          class="chakra-input__right-element css-dp9r5f"
        >
          <button
            aria-label="Search student"
            class="chakra-button css-15ql2fp"
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-gvh95d"
              focusable="false"
              viewBox="0 0 14 14"
            >
              <g
                clip-path="url(#clip0_63_207)"
                id="search 1"
              >
                <path
                  d="M13.9043 13.1687L10.377 9.64141C10.3141 9.57852 10.232 9.5457 10.1445 9.5457H9.86289C10.8008 8.53125 11.375 7.17773 11.375 5.6875C11.375 2.5457 8.8293 0 5.6875 0C2.5457 0 0 2.5457 0 5.6875C0 8.8293 2.5457 11.375 5.6875 11.375C7.17773 11.375 8.53125 10.8008 9.5457 9.86562V10.1445C9.5457 10.232 9.58125 10.3141 9.64141 10.377L13.1688 13.9043C13.2973 14.0328 13.5051 14.0328 13.6336 13.9043L13.9043 13.6336C14.0328 13.5051 14.0328 13.2973 13.9043 13.1687ZM5.6875 10.5C3.02695 10.5 0.875 8.34805 0.875 5.6875C0.875 3.02695 3.02695 0.875 5.6875 0.875C8.34805 0.875 10.5 3.02695 10.5 5.6875C10.5 8.34805 8.34805 10.5 5.6875 10.5Z"
                  fill="#C4C4C4"
                  id="Vector"
                />
              </g>
              <defs>
                <clippath
                  id="clip0_63_207"
                >
                  <rect
                    fill="white"
                    height="14"
                    width="14"
                  />
                </clippath>
              </defs>
            </svg>
          </button>
        </div>
      </form>
      <svg
        class="chakra-icon css-146j6cz"
        focusable="false"
        viewBox="0 0 17 20"
      >
        <g
          clip-path="url(#clip0_93_214)"
        >
          <path
            d="M8.50024 18.75C9.17038 18.75 9.71453 18.1883 9.71453 17.4988H10.9288C10.9288 18.8781 9.83899 20 8.50024 20C7.16149 20 6.07167 18.8781 6.07167 17.4988H7.28596C7.28596 18.1883 7.83011 18.75 8.50024 18.75ZM0.555017 13.0785C1.616 12.0387 2.42729 10.9504 2.42729 7.26172C2.42729 4.15273 4.83272 1.61719 7.8931 1.31055V0.625C7.8931 0.279687 8.1648 0 8.50024 0C8.83569 0 9.10739 0.279687 9.10739 0.625V1.31094C12.1678 1.61758 14.5732 4.15273 14.5732 7.26172C14.5732 10.95 15.3849 12.0387 16.4458 13.0785C16.976 13.598 17.1433 14.3836 16.8727 15.0797C16.5965 15.7906 15.9317 16.25 15.1788 16.25H1.82167C1.06881 16.25 0.40399 15.7902 0.127739 15.0793C-0.142818 14.3832 0.0249043 13.598 0.555017 13.0785ZM1.82167 15H15.1788C15.7188 15 15.9886 14.3566 15.6084 13.984C14.2856 12.6875 13.3589 11.2363 13.3589 7.26211C13.3589 4.63008 11.1857 2.5 8.50024 2.5C5.81515 2.5 3.64158 4.62969 3.64158 7.26172C3.64158 11.2207 2.7229 12.6797 1.39212 13.9836C1.01037 14.3582 1.28397 15 1.82167 15Z"
            fill="#C4C4C4"
          />
        </g>
        <defs>
          <clippath
            id="clip0_93_214"
          >
            <rect
              fill="white"
              height="20"
              transform="matrix(-1 0 0 1 17 0)"
              width="17"
            />
          </clippath>
        </defs>
      </svg>
    </div>
  </div>
</div>
`;
